### DISCUSSION TITLE
Add routes from one router into another programatically

### DISCUSSION THREAD

**Discussion by gerazenobi** [fastapi/fastapi#11136] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
for route in router_v1.routes:
    if route.path != "...":
       router_v2.routes.append(route)
```

### Description

Hi there 👋 

I was wondering how to import routes from 1 router into another one: I have a v1 router and I want the v2 router to have the same routes + others.

I quickly tested the following and it works:

```python
for route in router_v1.routes:
    if route.path != "...":
       router_v2.routes.append(route)
```

I wanted to know if it was safe to do so (e.g. `routes.append`) and whether I was not hacking or breaking any internals whatsoever.

### Operating System

macOS

### FastAPI Version

0.108.0

### Pydantic Version

1.10.13

### Python Version

Python 3.9.16

**Comment by DJoepie** [comment]
You could break up your routes into separate files and include them in your `main.py` as described [here](https://fastapi.tiangolo.com/tutorial/bigger-applications/#bigger-applications-multiple-files)
Or am I misunderstanding what you're trying to do?

I have:
![image](https://github.com/tiangolo/fastapi/assets/78362619/64d1679b-24c4-4f23-a36c-c4ef1033e5e4)

and then my `main.py` looks something like:
```python
from fastapi import FastAPI
from src.routes import invites, oauth, users

Base.metadata.create_all(bind=engine)

app = FastAPI()

# Include routes from different modules
app.include_router(oauth.router)
app.include_router(users.router)
app.include_router(invites.router)
```

**Reply by gerazenobi** [reply]
@DJoepie thanks for your quick reply.

My routes are already in separate modules.

```py
app.include_router(router_v1, prefix="/v1")
app.include_router(router_v2, prefix="/v2")
```

What I was doing is adding v1 routes into the v2 router programmatically  and wanted to confirm if my approach for doing so was OK and had no "side effects"

**Reply by DJoepie** [reply]
ahh fair enough, that's actually a good question and I'm personally not too sure if that would break something.
If it works and none of your tests fail, maybe it should be fine, however I'm not sure so maybe someone with some expertise could answer this question better.

**Comment by YuriiMotov** [comment, marked as answer]
You can just include router into the router:

```python
import fastapi

# Router 1

r1 = ApiRouter = fastapi.APIRouter()

@r1.get("/something")
async def something():
    return {"message": "This is 'something' of r1"}

@r1.get("/something-2")
async def something_2():
    return {"message": "This is 'something 2' of r1"}


# Router 2

r2 = ApiRouter = fastapi.APIRouter()

@r2.get("/something")
async def r2_something():
    return {"message": "This is 'something' of r2"}

@r2.get("/something-3")
async def r2_something_3():
    return {"message": "This is 'something 3' of r2"}

r2.include_router(r1)


# Main app

app = fastapi.FastAPI()

app.include_router(r1, prefix="/r1", tags=["r1"])

app.include_router(r2, prefix="/r2", tags=["r2"])

```

And if you want to exclude some of routes of router 1 to be accessed via router 2, you can re-define this route in router 2 (e.g. return 404 response). Note: it's important to add such routes before including router 1!
